### ISSUE TITLE
difference

### ISSUE THREAD

**Issue by GGyynnn** [ttgump/scDeepCluster#10]
What is the difference between scDeepCluster.py,scDeepCluster_latent.py and scDeepCluster_estimate_k.py?

**Comment by ttgump** [comment]
> What is the difference between scDeepCluster.py,scDeepCluster_latent.py and scDeepCluster_estimate_k.py?

"scDeepCluster.py" is the code of scDeepCluster model.
"scDeepCluster_latent.py" is the code used to extract latent embedding after the model training.
"scDeepCluster_estimate_k.py" is the code to use an intuitive metric generalizability to estimate the number of clusters from data.

**Comment by GGyynnn** [comment]
Thank you very much for your reply to my question, but I still have a doubt. The network structure diagram run out with the code you gave seems to be different from the network structure diagram you gave on github. The main difference is in the way of adding Gaussian noise. Attached below is the diagram you gave with a diagram of the code running.



---- Replied Message ----
| From | Tian ***@***.***> |
| Date | 03/24/2024 02:37 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [ttgump/scDeepCluster] difference (Issue #10) |

What is the difference between scDeepCluster.py,scDeepCluster_latent.py and scDeepCluster_estimate_k.py?

"scDeepCluster.py" is the code of scDeepCluster model.
"scDeepCluster_latent.py" is the code used to extract latent embedding after the model training.
"scDeepCluster_estimate_k.py" is the code to use an intuitive metric generalizability to estimate the number of clusters from data.

**Comment by ttgump** [comment]
Sorry I can't see your attachment, could you send it again? Thanks.

**Comment by GGyynnn** [comment]
Sorry, I sent it again, and I found that you may not be able to open the image link on github, you can try to view it directly in the mailbox. Thank you!



---- Replied Message ----
| From | Tian ***@***.***> |
| Date | 03/26/2024 01:01 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [ttgump/scDeepCluster] difference (Issue #10) |

Sorry I can't see your attachment, could you send it again? Thanks.

**Comment by ttgump** [comment]
It seems I didn’t receive your attachment. Could you send it to another email box: ***@***.******@***.***>. Thanks.
Best,
Tian

From: GGyynnn ***@***.***>
Date: Tuesday, March 26, 2024 at 9:50 AM
To: ttgump/scDeepCluster ***@***.***>
Cc: Tian Tian ***@***.***>, Comment ***@***.***>
Subject: Re: [ttgump/scDeepCluster] difference (Issue #10)
Sorry, I sent it again, and I found that you may not be able to open the image link on github, you can try to view it directly in the mailbox. Thank you!



---- Replied Message ----
| From | Tian ***@***.***> |
| Date | 03/26/2024 01:01 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [ttgump/scDeepCluster] difference (Issue #10) |

Sorry I can't see your attachment, could you send it again? Thanks.

**Comment by GGyynnn** [comment]
I am very sorry that this email cannot be displayed here, and the message was sent by github on behalf of me. Can you show your email address on github's author profile if it's convenient? Or can we communicate via wechat? My wechat is gyn496418_.(Maybe wechat can not be displayed in your place)

**Comment by GGyynnn** [comment]
抱歉，我知道加微信这件事很唐突。我发现github上可以上传图片。下面分别是您在github上给出的网络结构和我用您的代码跑出来的网络结构。主要区别似乎是高斯噪声的添加方式。请问这有影响吗？非常期待您的回答。
![1](https://github.com/ttgump/scDeepCluster/assets/150310503/d5a9dc3c-4d20-42a4-9ce8-17a387486190)
![2](https://github.com/ttgump/scDeepCluster/assets/150310503/11cd97e4-b408-420f-8724-c3165941d6fa)

**Comment by GGyynnn** [comment]
It's the network structure you gave on github and the network structure I ran out with your code. The main difference seems to be the way Gaussian noise is added. Does that make a difference? I am looking forward to your reply.
![1](https://github.com/ttgump/scDeepCluster/assets/150310503/71d67919-1c4b-4296-992a-d2d46f8362dc)
![2](https://github.com/ttgump/scDeepCluster/assets/150310503/8d19d0a9-15b5-4b6b-9d83-3f6513232939)

**Comment by GGyynnn** [comment]
I'm sorry. I know it was a little abrupt. I found that github can upload pictures. The following are the network structure you gave on github and the network structure I ran out with your code. The main difference seems to be the way Gaussian noise is added. Does that make a difference? I am looking forward to your reply.
![1](https://github.com/ttgump/scDeepCluster/assets/150310503/e8870932-cb0c-44e8-8813-c09311d6182a)
![2](https://github.com/ttgump/scDeepCluster/assets/150310503/bd884688-b633-406a-bfff-f1c8afca80b4)

**Comment by ttgump** [comment]
Yes, there is difference. The main difference is I didn't add noise to the clustering layer, which will make the model difficult to converge.

**Comment by GGyynnn** [comment]
I'm sorry, I don't understand what you mean. You said that you did not add noise to the clustering layer, but because I modified the network to remove the clustering layer. The way I'm talking about noise being added differently is during encoding . As shown in the following figure. Also thanks for your advice, I will try the PyTorch version. Can you answer the questions I have raised about this code? What causes these two different ways of adding noise? Is it the structure of the network or the keras version? Do different noise adding methods have a big impact on the overall network?
![111](https://github.com/ttgump/scDeepCluster/assets/150310503/3ed7d187-3fbc-46ef-955f-63a9034d96a6)
